### PR TITLE
feat(MeshCNN): Add MeshCNN's Convolution Layer

### PR DESCRIPTION
**Pull request by michaelfortunato**
This commit is the first part of a series of pull requests that we will make to PyTorch Geometric in order add MeshCNN! MeshCNN is a complex and important paper. As such its integration into PyG will be via a series of self-contained PRs. The proposed plan to integrate MeshCNN into PyTorch Geometric is posted as a new comment on the original issue: #283. The full PDF is also downloadable here: https://github.com/user-attachments/files/19872710/MeshCNN-PyG.pdf.

# Background

It's important to recognize that the original attempt to integrate MeshCNN into PyG, PR #2293, is quite comprehensive and complete. The code we put here for MeshCNN will may very well look like the original PR. The main difference is that we will split up the pull requests into a series of smaller ones.

This is Part 1 of a comprehensive plan to add MeshCNN into PyTorch Geometric and implements the convolutional layer of MeshCNN. Please see #283 for the comprehensive plan. Please add all comments and feedback related to the entire plan or higher level ideas on the original issue #283. I would love critical feedback on the plan and am happy to completely throw it away or change it based on the feedback I get. 

# PR's Design

A good place to fully-understand `MeshCNNConv`'s design is by rendering the docs for `torch.nn.conv.MeshCNNConv` and reading it.

The most important aspect of this pull request is ordering we require on the `edge_index` (edge adjacency). This constraint is documented in the class (`torch.nn.conv.MeshCNNConv`).

My questions are:

1. Should I write the core logic of MeshCNN's (non-permutation invariant) convolution as a new `Aggregate` module?
2. MeshCNN's convolution can actually be expressed in terms of the framework of `torch_geometric.nn.conv.RGCNConv`, where in MeshCNN there are 4 categories. Should we sub-class that instead?
3. Should we implement batch normalization as a performance optimization here?

I have thought about these questions a lot, and if any of these questions are "yes", my rebuttals to them are

1. MeshCNN's aggregation is not generalizable enough to warrant the creation of a new aggregation file. That would introduce two new files to implement `MeshCNNConv`, that is `torch.nn.aggr.MeshCNNConv`, and the smaller wrapper `torch.nn.conv.MeshCNNConv`. That being said, moving to `torch.nn.aggr.MeshCNNConv` is easy to do.
2. While MeshCNN can be expressed in terms of `torch_geometric.nn.conv.RGCNConv`, it ties us to any performance characteristics etc that this `torch_geometric.nn.conv.RGCNConv` may have. I think it is wrong to couple us to this framework, even if conceptually it is important and interesting to recognize MeshCNNConv as a case of `torch_geometric.nn.conv.RGCNConv`.
3. This can come as a future and enhancement PR after MeshCNN is more situated into PyG. Adding this complexity now obscures the main and lasting design decisions around our implementation of MeshCNN. These first PR's should get the core functionality out in a way that is straightforward, which will increase the odds that our design decisions now can be easily extended to accommodate future enhancements.

Besides that, this is a small PR, most of the lines being documentation.

Thanks. _I am a student in Chicago working independently on this project._ **As of this week, I am working on this project full-time.**

> [!IMPORTANT] 
> I am always reachable directly via instant message or video calling in PyG's slack. My display name is `@Michael Fortunato` and my account link is https://torchgeometricco.slack.com/team/U08AMAGE318. Please feel free to message or call me directly, at anytime, with questions or comments. I am also always happy to meet if that helps get things moving.

Thanks so much. Once again, for background on the entire plan please see my comments on
the original issue #283 along with the attached PDF of the plan. To get started with this PR, I recommend viewing the rendered class documentation for `torch.nn.conv.MeshCNNConv`: https://pytorch-geometric--10222.org.readthedocs.build/en/10222/generated/torch_geometric.nn.conv.MeshCNNConv.html.

Best,
Michael

Slack: `@Michael Fortunato` on [Slack](https://torchgeometricco.slack.com/team/U08AMAGE318)
Original Issue Along With Plan Summary: #283
Full Proposal: [MeshCNN-PyG.pdf](https://github.com/user-attachments/files/19873871/MeshCNN-PyG.pdf)

### CODE REVIEW

File: torch_geometric/nn/conv/meshcnn_conv.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+        self.out_channels = out_channels
+
+        if kernels is None:
+            self.kernels = ModuleList(
+                [Linear(in_channels, out_channels) for _ in range(5)])
+
+        else:
+            # ensures kernels is properly formed, otherwise throws
+            # the appropriate error.
+            self.__assert_kernels(kernels)
```
Comment (rusty1s) [inline torch_geometric/nn/conv/meshcnn_conv.py]:
```suggestion
            self._assert_kernels(kernels)
```

File: torch_geometric/nn/conv/meshcnn_conv.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+            returned by aggregate.
+           x(torch.Tensor): :math:`X^{(k)}`. The original inputs to this layer.
+
+        Returns:
+            torch.Tensor: :math:`X^{(k+1)}`. The output of this layer, which
+            has shape :obj:`(|E|, out_channels)`.
+        """
+        return self.kernels[0].forward(x) + inputs
+
+    def __assert_kernels(self, kernels: ModuleList):
```
Comment (rusty1s) [inline torch_geometric/nn/conv/meshcnn_conv.py]:
```suggestion
    def _assert_kernels(self, kernels: ModuleList):
```

File: torch_geometric/nn/conv/meshcnn_conv.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+        # m = torch.stack(
+        #     [
+        #         (n_a - n_c).abs(),  # shape: |E| x in_channels
+        #         n_a + n_c,
+        #         (n_b - n_d).abs(),
+        #         n_b + n_d,
+        #     ],
+        #     dim=1)  # shape: (|E| x 4 x in_channels)
+        # m.view(E4, in_channels)  # shape 4*|E| x in_channels
+        # return m
```
Comment (rusty1s) [inline torch_geometric/nn/conv/meshcnn_conv.py:416]:
```suggestion
        # return m

```

**Review by rusty1s — APPROVED** [review]

**Comment by rusty1s** [comment]
Very in-depth documentation, love it :) Thank you so much.
